### PR TITLE
color, opacity and radius legends

### LINKED ISSUE #23: Color legend.
When using a color legend (e.g., #2 #19), should we automatically include a nice [legend](https://observablehq.com/@d3/color-legend) (probably yes)?

### LINKED ISSUE #236: radius legend?
This might be a bit difficult to do (do we want superposed circles? how to chose the reference values? even by hand it's difficult…), and it would in any case need inferRadialRange (#235) to be feasible.

related #23

### PR DESCRIPTION
**Pull request by Fil**
sequel of #425

build at https://observablehq.com/d/c324a9a466e66329

todo:
- [x] fix API (1 options object)
- [x] examples
- [x] remove dependency on htl
- [x] fix the figure so it gets the same width as the plot
- [x] expose and document Plot.colorLegend
- [x] continuous date color legend

closes #23 
closes #236

related: #397 #92

### CODE REVIEW

**Comment by mbostock** [comment]
The color legend should inherit the scale’s automatic label, no? E.g., here both the _x_-axis and color legend should be labeled with “random”, without you needing to specify a label explicitly.

```js
Plot.plot({
  color: {
    legend: true,
    width: 120,
    nice: true
  },
  marks: [
    Plot.dot({length: 100}, {
      x: Object.assign(() => Math.random(), {label: "random"}),
      fill: Object.assign(() => Math.random(), {label: "random"})
    })
  ]
})
```

File: src/legends/color.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  }
+  return legend(scale(descriptor), {
+    title: descriptor.label,...options
+  });
+}
+
+function legend(color, {
+  title,
+  tickSize = 6,
+  width = 320,
```
Comment (mbostock) [inline src/legends/color.js:19]:
```suggestion
  width = 240, 
```

320 feels a little wide as a default. Should probably also be capped at the width of the chart in the legend: true case.

File: src/scales.js [hunk 2]
```diff
@@ -116,25 +119,34 @@ function asOrdinalType(key) {
...
+  const scales = {};
+  const legends = [];
   for (const key in scaleDescriptors) {
     let cache;
     Object.defineProperty(scales, key, {
       enumerable: true,
       get: () => cache = cache || exposeScale(scaleDescriptors[key])
     });
+    const options = scaleDescriptors[key].options;
+    let {legend} = options;
```
Comment (mbostock) [inline src/scales.js:132]: This is scoping the legend options under a legend object, which is different from how we handle axis options (which are in the same namespace as scale options, i.e., not under an axis object). I wonder whether we should flatten the legend options as well; certainly the ticks and label options seem the same as an axis, so no reason to make them different.

File: src/scales.js [hunk 3]
```diff
@@ -116,25 +119,34 @@ function asOrdinalType(key) {
   return registry.get(key) === position ? "point" : "ordinal";
 }
 
-export function exposeScales(figure, scaleDescriptors) {
-  const scales = figure.scales = {};
+export function exposeScales(scaleDescriptors) {
```
Comment (mbostock) [inline src/scales.js:122]: The name “exposeScales” no longer seems accurate. Can we separate the two functions, keeping exposeScales as it was and having a separate method for computing the legends?

File: src/legends/swatches.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  max-width: calc(100% - ${+swatchWidth}px - 0.5em);
+}
+
+.${id}-swatch {
+  width: ${+swatchWidth}px;
+  height: ${+swatchHeight}px;
+  margin: 0 0.5em 0 0;
+}
+
+  </style>
```
Comment (mbostock) [inline src/legends/swatches.js:37]: We can move all of this to style.css and then just use CSS variables for the variables (the swatch width, height, and color). Then we won’t need uid or inline styles (other than to define the CSS variables), either.

**Comment by Fil** [comment]
> color legend should inherit the scale’s automatic label

Yes that's the intent. I swear it was working at some point :)

Thank you for the review! I should have marked this as a Draft because there's quite a lot of screws to tighten.

I think this shows that the "scale options" object (returned as plot.scales.color etc) should contain all the options that we passed and that were not consumed when building the scale.

Also, I want more unit tests—for example I realize that calling Plot.scale on a diverging scale will crash if we don't tell it that the key is "color".

**Comment by Fil** [comment]
deferring to #484

**Comment by Fil** [comment]
see https://github.com/observablehq/plot/pull/665

**Pull request closed without merging**